### ISSUE TITLE
Ansible tasks slow down as host facts grow

### ISSUE THREAD

**Issue by mwhahaha** [ansible/ansible#73654]
##### SUMMARY
As the size of host facts is increased, there is a serious slow down in the number of tasks that can be processed per second.  This can be viewed at smaller scale, but is dramatic at ~150 nodes.  This can happen because of network facts (with many interfaces 100+), service facts, or package facts. This can also be easily reproduced with loading a large static fact structure (>100KB) and filtering to ansible_local.

I was able to pin point one of the issues coming from the get_vars in the host task processing:
https://github.com/ansible/ansible/blob/stable-2.9/lib/ansible/plugins/strategy/linear.py#L282-L283

This was seen to take <10ms with no facts and ~200ms with facts.  Since host task processing is performed serially, the result is that <= 5 tasks can be processed per second at this scale.

There is another issue with the amount of memory and the workers being spawned, but the task execution slowdown seems to be more consistent.

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME

lib/ansible/vars/manager.py
lib/ansible/plugins/strategy/linear.py

##### ANSIBLE VERSION

ansible-2.9.17-1.el8ae.noarch
```paste below
$ ansible --version
ansible 2.9.17
  config file = /home/stack/scale-ssh-ansible/ansible.cfg
  configured module search path = ['/home/stack/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python3.6/site-packages/ansible
  executable location = /usr/bin/ansible
  python version = 3.6.8 (default, Oct 11 2019, 15:04:54) [GCC 8.3.1 20190507 (Red Hat 8.3.1-4)]
```

##### CONFIGURATION

```paste below
ANSIBLE_PIPELINING(/home/stack/scale-ssh-ansible/ansible.cfg) = True
ANSIBLE_SSH_ARGS(/home/stack/scale-ssh-ansible/ansible.cfg) = -o UserKnownHostsFile=/dev/null -o StrictHostKey>
ANSIBLE_SSH_CONTROL_PATH_DIR(/home/stack/scale-ssh-ansible/ansible.cfg) = /tmp/scale-ansible-ssh
ANSIBLE_SSH_RETRIES(/home/stack/scale-ssh-ansible/ansible.cfg) = 8
DEFAULT_CALLBACK_WHITELIST(/home/stack/scale-ssh-ansible/ansible.cfg) = ['profile_tasks']
DEFAULT_FORKS(/home/stack/scale-ssh-ansible/ansible.cfg) = 32
DEFAULT_INTERNAL_POLL_INTERVAL(/home/stack/scale-ssh-ansible/ansible.cfg) = 0.005
DEFAULT_TIMEOUT(/home/stack/scale-ssh-ansible/ansible.cfg) = 30
INTERPRETER_PYTHON(/home/stack/scale-ssh-ansible/ansible.cfg) = auto
```

##### OS / ENVIRONMENT

RHEL8.2

This has been seen on OpenStack compute nodes with an average of 200 network interfaces. This same effect can be reproduced without network facts and using service facts or package facts.

##### STEPS TO REPRODUCE

```yaml
# baseline.yml
- hosts: all
  gather_facts: false
  tasks:
    - meta: clear_facts
    - name: Gather facts
      setup:
        gather_subset:
          - '!all'
          - min
    - name: Sleep 1
      shell: sleep 1
```
```yaml
# package-facts.yml
- hosts: all
  gather_facts: false
  tasks:
    - meta: clear_facts
    - name: Gather facts
      setup:
        gather_subset:
          - '!all'
          - min
    - name: Gather package facts
      package_facts:
        manager: auto
    - name: Sleep 1
      shell: sleep 1
```
```yaml
# service-facts.yml
- hosts: all
  gather_facts: false
  tasks:
    - meta: clear_facts
    - name: Gather facts
      setup:
        gather_subset:
          - '!all'
          - min
    - name: Gather service facts
      service_facts: # sup
    - name: Sleep 1
      shell: sleep 1
```

##### EXPECTED RESULTS

Task execution should not slow down when the number of Ansible host facts increase.

##### ACTUAL RESULTS

Execution against 179 hosts

```paste below
$ ansible-playbook -i inventory.ini baseline.yml
...SNIP...
Thursday 18 February 2021  15:32:19 +0000 (0:00:04.784)       0:00:08.455 ***** 
=============================================================================== 
Sleep 1 ----------------------------------------------------------------- 4.78s
Gather facts ------------------------------------------------------------ 3.58s
```
```
$ ansible-playbook -i inventory.ini package-facts.yml
...SNIP...
Thursday 18 February 2021  15:34:26 +0000 (0:00:26.821)       0:00:52.222 ***** 
=============================================================================== 
Sleep 1 ---------------------------------------------------------------- 26.82s
Gather package facts --------------------------------------------------- 21.53s
Gather facts ------------------------------------------------------------ 3.78s
```
```
$ ansible-playbook -i inventory.ini service-facts.yml
...SNIP...
Thursday 18 February 2021  15:36:00 +0000 (0:00:08.965)       0:00:26.258 ***** 
=============================================================================== 
Gather service facts --------------------------------------------------- 12.44s
Sleep 1 ----------------------------------------------------------------- 8.97s
Gather facts ------------------------------------------------------------ 3.94s
```

**Comment by ansibot** [comment]
Files identified in the description:
* [`lib/ansible/plugins/strategy/linear.py`](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/strategy/linear.py)
* [`lib/ansible/vars/manager.py`](https://github.com/ansible/ansible/blob/devel/lib/ansible/vars/manager.py)

If these files are incorrect, please update the `component name` section of the description or use the `!component` bot command.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by samdoran** [comment]
This is most likely due to `INJECT_FACTS_AS_VARS`, which is currently set to `True` by default. Since all the facts are reevaluated and when set to vars, this can be quite slow if there is a large amount of data to process. Eliminating this extra step may speed things up considerably.

Another thing that could help is changing the stage at which variables are evaluated for the [`host_group_vars`](https://docs.ansible.com/ansible/latest/collections/ansible/builtin/host_group_vars_vars.html) inventory plugin. By default, `host_group_vars` is run for every task ("demand"). Try setting `ANSIBLE_VARS_PLUGIN_STAGE` to "inventory" to prevent vars from being reevaluated for every task.

**Comment by mwhahaha** [comment]
Looks like the plugin stage configuration per https://docs.ansible.com/ansible/latest/plugins/vars.html#using-vars-plugins is:

```
[vars_host_group_vars]
stage = inventory
```

This didn't seem to improve anything.

```
$ grep vars_host_group_vars ansible.cfg -A1
[vars_host_group_vars]
stage = inventory
$ time ansible-playbook -i tripleo-ansible-inventory.yaml baseline.yml
...SNIP...
Thursday 18 February 2021  20:17:21 +0000 (0:00:05.508)       0:00:08.781 ***** 
=============================================================================== 
Sleep 1 ----------------------------------------------------------------- 5.51s
Gather facts ------------------------------------------------------------ 3.22s

real	0m9.964s
user	0m11.818s
sys	0m10.913s

$ time ansible-playbook -i tripleo-ansible-inventory.yaml package-facts.yml
...SNIP...
Thursday 18 February 2021  20:18:21 +0000 (0:00:25.774)       0:00:47.194 ***** 
=============================================================================== 
Sleep 1 ---------------------------------------------------------------- 25.77s
Gather package facts --------------------------------------------------- 18.13s
Gather facts ------------------------------------------------------------ 3.24s

real	0m48.969s
user	1m32.909s
sys	0m22.167s
```

`ANSIBLE_INJECT_FACT_VARS` does improve it a bit, but still a significant slow down.
```
$ export ANSIBLE_INJECT_FACT_VARS=False
$ time ansible-playbook -i tripleo-ansible-inventory.yaml baseline.yml
...SNIP...
Thursday 18 February 2021  20:11:48 +0000 (0:00:05.635)       0:00:08.956 ***** 
=============================================================================== 
Sleep 1 ----------------------------------------------------------------- 5.64s
Gather facts ------------------------------------------------------------ 3.27s

real	0m10.098s
user	0m11.356s
sys	0m10.951s

$ time ansible-playbook -i tripleo-ansible-inventory.yaml package-facts.yml
...SNIP...
Thursday 18 February 2021  20:12:41 +0000 (0:00:13.446)       0:00:35.216 ***** 
=============================================================================== 
Gather package facts --------------------------------------------------- 18.65s
Sleep 1 ---------------------------------------------------------------- 13.45s
Gather facts -------------------------------------------------------------3.06s

real	0m37.053s
user	1m11.846s
sys	0m22.419s
```

**Comment by samdoran** [comment]
That's good that disable fact inject made a difference.

Ultimately, I'm not sure there is much we can do without significantly rewriting how we process the data. More data takes more time to process, and we process it _a lot_.

I am working on some things for the 2.11 release right now. I will dig into this again once that work is complete.

**Comment by mwhahaha** [comment]
Sure. Ping me when you have time and I can share what I've found. Package facts brings the fact size (at least in cache) up from 5.5k to ~313k so we're dealing with 1.6M vs 58M of facts.  Network facts for ended up being like 123M

**Comment by egmar** [comment]
I'm seeing similar behavior and is mostly memory which worries us. After doing some investigation with the help of `memory_profiler` in `VariableManager.get_vars` method
```
mem_usage = memory_usage(-1, interval=.2, timeout=1)
display.warning(mem_usage)
```
the memory growth is quite noticeable
```
[WARNING]: [48.19921875, 48.19921875, 48.19921875, 48.19921875, 48.19921875]
[WARNING]: [48.2265625, 48.2265625, 48.2265625, 48.2265625, 48.2265625]

PLAY [test] ********************************************************************
[WARNING]: [48.49609375, 48.49609375, 48.49609375, 48.49609375, 48.49609375]
[WARNING]: [61.25390625, 61.25390625, 61.25390625, 61.25390625, 61.25390625]
[WARNING]: [71.4296875, 71.484375, 71.484375, 71.484375, 71.484375]
[WARNING]: [82.0390625, 82.0390625, 82.0390625, 82.0390625, 82.0390625]
[WARNING]: [92.5546875, 92.5546875, 92.5546875, 92.5546875, 92.5546875]
[WARNING]: [103.05078125, 103.05078125, 103.05078125, 103.05078125,
103.05078125]
[WARNING]: [113.7734375, 113.7734375, 113.7734375, 113.7734375, 113.7734375]
[WARNING]: [124.0078125, 124.0078125, 124.0078125, 124.0078125, 124.0078125]
[WARNING]: [134.8359375, 134.8359375, 134.8359375, 134.8359375, 134.8359375]
[WARNING]: [145.2578125, 145.2578125, 145.2578125, 145.2578125, 145.2578125]
[WARNING]: [156.21484375, 156.21484375, 156.21484375, 156.21484375,
```

I also noticed that a lot of memory is consumed by AnsibleUnsafe objects (it seems Ansible is converting all facts, both keys and values to be of `AnsibleUnsafe` type. I used `pympler` in same place to capture a summary of memory used by all objects with following results (this is a snapshot, it generally increases over time as playbook runs):
```
                                         types |   # objects |   total size
============================================== | =========== | ============
  ansible.utils.unsafe_proxy.AnsibleUnsafeText |      343564 |     35.34 MB
                                          dict |      141330 |     33.67 MB
                                           str |       35563 |      3.64 MB
                                          code |       11488 |      1.96 MB
                                          type |        1727 |      1.57 MB
                                          list |        5090 |      1.45 MB
                                         tuple |        9473 |    523.06 KB
                                   abc.ABCMeta |         349 |    359.41 KB
                                           set |         348 |    234.16 KB
   ansible.parsing.yaml.objects.AnsibleUnicode |        1980 |    218.72 KB
                                       weakref |        2775 |    195.12 KB
                    builtin_function_or_method |        2482 |    174.52 KB
                                     frozenset |         447 |    152.04 KB
                                           int |        4223 |    127.19 KB
   ansible.parsing.yaml.objects.AnsibleMapping |         459 |    117.28 KB
```

This is with each host in inventory having ~1MB of facts and `strategy: free`. 
I'm also attaching the call graph (perhaps will be helpful) which I collected by running the playbook with PyCharm profiler and `yappi`
![ansible3](https://user-images.githubusercontent.com/243891/108963054-e6c93b00-7679-11eb-8ca0-ea53ef346604.png)

**Comment by egmar** [comment]
Small update, I've run some more tests and it could be pointed that facts content (keys and values) wrapping into `AnsibleUnsafeText` is mostly the cause of increased memory consumption. 
https://github.com/ansible/ansible/blob/eb72c36a71c8bf786d575a31246f602ad69cc9c9/lib/ansible/utils/unsafe_proxy.py#L135-L136

**Comment by mykaul** [comment]
> This is most likely due to `INJECT_FACTS_AS_VARS`, which is currently set to `True` by default. Since all the facts are reevaluated when injecting them as vars, this can be quite slow if there is a large amount of data to process. Eliminating this extra step may speed things up considerably.
> 
> Another thing that could help is changing the stage at which variables are evaluated for the [`host_group_vars`](https://docs.ansible.com/ansible/latest/collections/ansible/builtin/host_group_vars_vars.html) inventory plugin. By default, `host_group_vars` is run for every task ("demand"). Try setting `ANSIBLE_VARS_PLUGIN_STAGE` to "inventory" to prevent vars from being reevaluated for every task.

Would be nice to add a comment to https://docs.ansible.com/ansible/latest/reference_appendices/config.html#inject-facts-as-vars with the above information and alternative implementation.

**Comment by freedge** [comment]
> There is another issue with the amount of memory and the workers being spawned, but the task execution slowdown seems to be more consistent.

Some observation  - I'm measuring here the clone system call taking 50ms (likely due to the 1.5GB RSS ansible process). The time for the clone seems to increase linearly with the RSS.

To measure that we can artificially augment the RSS of the ansible process in a dummy playbook by creating a fact with something like this at the beginning

```
      - set_fact:
          zeroes: "{{lookup('file', 'zeroes') }}"
        run_once: yes
        delegate_to: localhost
```

where zeroes is a huge file we create before hand...

We also run the playbook once and limit to a single host, to be able to substract the cost of our extra task and be able to time only the effect on the rest of the tasks.

We also do another run, with ```strace -T -e clone ansible-playbook..```, to be able to time the clone system call for only the main process. Results look like that:

---

For 1 host with zeroes file of 250MB (cost of the extra task + playbook for 1 task)
```
real	0m10,638s
user	0m5,971s
sys	0m4,930s
```

For 200 hosts with zeroes file of 250MB (cost of that extra task + playbook running on 200 hosts)
```
real	0m43,107s
user	1m53,379s
sys	4m31,848s
```
with 16s spent in clone. RSS observed is ~1GB RSS. With the set_fact deducted it runs in 32s.

---

For 1 host with a file of 500MB (cost of the extra task + playbook limited to 1 host)
```
real	0m19,997s
user	0m11,795s
sys	0m8,672s
```

For 200 hosts with a file of 500MB (cost of that extra task + playbook running on 200 hosts)
```
real	1m9,116s
user	2m3,070s
sys	7m4,575s
```
with 27s spent in clone. RSS observed is ~2GB RSS. With the set_fact deducted it runs in 50s.


---

I conclude that the Ansible tasks slow down generally as memory grows :) , and this one system call alone accounts for half of the slowness. I guess if you worked already on reducing the memory it should logically already address this issue.

The workers are spawned once for each task, cloning the main process, even if the task is skipped (a bit enraging). For 200 hosts, and 500 tasks, and 50ms per clone, I estimate we spend more than 1h in the clone system call alone, for a 5h30 OpenStack deployment.

(also I believe this issue could be the underlying issue for https://github.com/ceph/ceph-ansible/issues/3021 - where ceph-ansible acts "as if it was running with serial: 1")

**Comment by bcoca** [comment]
The fork happens because we template at the 'loop level' and  the skip is per loop item, so we won't even know if the task has a loop or if it is skipped until much later in the process. 

We have been looking at 'pre-computing' this in the main thread but requires quite a re-architecture and will not be easy to accomplish while still maintaining the features and flexibility that the current model has. 

Doing the computation at the 'last moment' does allow us to save ourselves a lot of work in some cases and speeds up some scenarios. As you note adding a large memory footprint that we need to copy to each fork, is not one of them.

**Comment by phemmer** [comment]
In my case on this, the `wrap_var()` is accounting for about 35% of the CPU time. The `wrap_var()` call should be returning early if the var is already wrapped as `AnsibleUnsafe`. I would think it shouldn't be too difficult to just ensure the fact cache loader is always using `AnsibleUnsafe` vars on load.

I switched to using pickle for fact caching, and it helped this issue somewhat, as some of the variables are stored already wrapped with `AnsibleUnsafeText`, but not all of them.  
Plus there's no `AnsibleUnsafeDict`, `AnsibleUnsafeList`, etc. So it still has to traverse the entire structure to make sure every leaf is wrapped.

**Comment by sivel** [comment]
For 2.12 we are going to be working on a new proxy object, that we plan to replace the unsafe functionality with, and part of the goal is that we just wrap the container and not every key/value, or element within.  I'm not sure that replacing unsafe will happen in 2.12, since we have to build the generic proxy functionality first, but potentially by 2.13.

**Comment by egmar** [comment]
Perhaps another thing to do, is to use Python 3.8 [SharedMemoryManager](https://docs.python.org/3/library/multiprocessing.shared_memory.html) and afaik, Ansible 2.12 will require Python 3.8 anyway. Since Ansible forks generally target specific host, one place it could help is supplying only a sub-set of facts or vars in general, which should result in forks using less memory.

**Comment by Vladimir-csp** [comment]
I would also like to point out that invocations of `role.get_default_vars()` and ` role.get_vars()` from `ansible/vars/manager.py` reduce performance of playbooks exponentially, if role dependencies organized hierarchically. 
This happens because invocations of `get_default_vars()` and `get_vars()` from `ansible/playbook/role/__init__.py` for high-level roles are multiplied by **undeduplicated** dependencies for each lower level role.
This results in a significant pause before every task on every host starts, even skipped ones, almost nullifying parallelism, because it takes longer to prepare vars for a task on the next host than it takes to finish the task on previous host.

**Comment by Vladimir-csp** [comment]
Here is how this problem manifests.

(I just took a production playbook structure and swapped out role names to role_[1:35])
<details>
  <summary>Role Dependency Structure</summary>

```
role_1 depends: []
role_2 depends: []
role_3 depends: []

role_4 depends:
  - role_1
  - role_3

role_5 depends:
  - role_1
  - role_4

role_6 depends: []

role_7 depends:
  - role_1

role_8 depends:
  - role_1
  - role_4

role_9 depends:
  - role_1
  - role_8

role_10 depends:
  - role_1

role_11 depends:
  - role_1
  - role_7
  - role_9

role_12 depends:
  - role_1
  - role_7
  - role_9
  - role_11

role_13 depends:
  - role_1
  - role_9

role_14 depends:
  - role_1

role_15 depends:
  - role_1

role_16 depends:
  - role_1
  - role_9

role_17 depends:
  - role_1

role_18 depends:
  - role_1
  - role_9

role_19 depends:
  - role_9
  - role_3
  - role_4
  - role_18

role_20 depends:
  - role_9
  - role_18
  - role_3
  - role_4

role_21 depends:
  - role_9
  - role_18
  - role_4

role_22 depends: []

role_23 depends: []

role_24 depends:
  - role_1
  - role_7
  - role_9
  - role_17
  - role_15

role_25 depends: []

role_26 depends:
  - role_1
  - role_9
  - role_11
  - role_12
  - role_16

role_27 depends:
  - role_26
  - role_4

role_28 depends:
  - role_12

role_29 depends:
  - role_17
  - role_25
  - role_27
  - role_28

role_30 depends:
  - role_4

role_31 depends: []

role_32 depends:
  - role_29
  - role_30
  - role_3
  - role_31

role_33 depends: []

role_34 depends:
  - role_1
  - role_2
  - role_5
  - role_7
  - role_8
  - role_9
  - role_10
  - role_11
  - role_12
  - role_13
  - role_14
  - role_15
  - role_16
  - role_35
  - role_24
  - role_36

role_35 depends:
  - role_19
  - role_20
  - role_21
  - role_22
  - role_23

```
</details>

Running a playbook that lists:
```
roles:
  - role_34
  - role_32
  - role_33
```

expecting get_vars to go through:
```
role_1
role_2
role_3
role_4
role_5
role_6
role_7
role_8
role_9
role_10
role_11
role_12
role_13
role_14
role_15
role_16
role_17
role_18
role_19
role_20
role_21
role_22
role_23
role_24
role_25
role_26
role_27
role_28
role_29
role_30
role_31
role_32
role_33
```

But instead it goes through this:
<details>
  <summary>undeduplicated roles</summary>

```
role_1
role_2
role_1
role_3
role_4
role_5
role_6
role_5
role_5
role_1
role_7
role_1
role_3
role_4
role_8
role_1
role_3
role_4
role_8
role_9
role_1
role_10
role_1
role_7
role_1
role_3
role_4
role_8
role_9
role_11
role_6
role_11
role_1
role_7
role_1
role_3
role_4
role_8
role_9
role_1
role_7
role_1
role_3
role_4
role_8
role_9
role_11
role_6
role_11
role_12
role_1
role_3
role_4
role_8
role_9
role_13
role_1
role_14
role_1
role_15
role_1
role_3
role_4
role_8
role_9
role_16
role_1
role_3
role_4
role_8
role_9
role_1
role_17
role_1
role_3
role_4
role_8
role_9
role_18
role_1
role_3
role_4
role_8
role_9
role_3
role_1
role_3
role_4
role_1
role_3
role_4
role_8
role_9
role_18
role_19
role_1
role_3
role_4
role_8
role_9
role_1
role_3
role_4
role_8
role_9
role_18
role_3
role_1
role_3
role_4
role_20
role_1
role_3
role_4
role_8
role_9
role_1
role_3
role_4
role_8
role_9
role_18
role_1
role_3
role_4
role_21
role_22
role_23
role_1
role_7
role_1
role_3
role_4
role_8
role_9
role_1
role_17
role_1
role_15
role_24
role_1
role_3
role_4
role_8
role_1
role_17
role_25
role_1
role_3
role_4
role_8
role_9
role_1
role_7
role_1
role_3
role_4
role_8
role_9
role_11
role_6
role_11
role_1
role_7
role_1
role_3
role_4
role_8
role_9
role_1
role_7
role_1
role_3
role_4
role_8
role_9
role_11
role_6
role_11
role_12
role_1
role_3
role_4
role_8
role_9
role_16
role_26
role_1
role_3
role_4
role_27
role_1
role_7
role_1
role_3
role_4
role_8
role_9
role_1
role_7
role_1
role_3
role_4
role_8
role_9
role_11
role_6
role_11
role_12
role_28
role_29
role_1
role_3
role_4
role_30
role_3
role_31
role_32
role_33
```
</details>

**Comment by Vladimir-csp** [comment]
I started testing transition from 2.9 to 2.11, the situation got much worse.
After completing `role_1`, ansible seemingly hangs. Debug reveals that it starts very slowly to run iterations of get_vars.
At the end of role_1 messages were printed to the log for each host `META: role_complete for hostname`, after that a large amount of empty `META:` messages started pouring.
In debug output a lot of stuff related to getting vars and building task lists seemingly on a loop:
```
getting the next task for host hostname
done getting next task for host hostname
 ^ task is: TASK: some role : some task name
Calling all_inventory to load vars for hostname
Calling groups_inventory to load vars for hostname
Calling all_plugins_inventory to load vars for hostname
Calling all_plugins_play to load vars for hostname
Calling groups_plugins_inventory to load vars for hostname
Calling groups_plugins_play to load vars for hostname
Loading data from /...host_vars/hostname

```
I am assuming it slowly crawls over the undeduplicated list of roles, but it is taking way too long. It took 10 minutes to complete. `profile_tasks` callback counted those 10 minutes as running time of the last task of `role_1` As a playbook was launched with a non-existing tag, only tasks with `always` tag were run, which are the entirety of `role_1`.

**Comment by ursetto** [comment]
This has been an ongoing problem over the last 3 years in our environment of >5000 servers.

My current workaround is to use `serial: 250` or lower, and `meta: clear_facts` at the *end* of the play, to clear all facts every 250 hosts. This prevents the seemingly exponential growth in runtime and memory from getting out of hand. `gather_subset: min` might suffice, but anything more (like `package_facts` or a full gather) will cause Ansible to fall over on deployments of any reasonable size (say, >1000), unless you aggressively prune facts.

The second workaround is to use job slicing if you have awx, which will divide up your effective inventory. (But never ever use a global fact cache.)

setup + filter doesn't work in practice because it only supports a single shell-style wildcard. If you need only `ansible_distribution` (from `min`) and `ansible_virtualization_role` (outside `min`) you have to gather everything.

So clearing everything occasionally is the only reliable solution. In general I only use facts from the current system, so keeping everything around is not necessary. Unfortunately using `serial` doesn't fit in to every playbook.

It would be ideal to address this at the source so Ansible can work in medium to large deployments without these workarounds.

**Comment by samdoran** [comment]
I have been moved to another team and will unfortunately not be able to look into this. 😞

**Comment by ryanmerolle** [comment]
Hopefully not too off topic, but I’ve been on a recent kick to optimize some plays and workflows I have, and as hostvars increase I do see a correlation to runtimes increase especially when run against large host inventories.  I’m going to test this, but if I use `set_facts` just before I need variables to be defined, and `clear_facts` after I no longer need said variables in a play, would that help in resource usage and runtime?

This should not be too far off topic given I’m trying to workaround the same root cause of slowdowns.  I was just not clear if `clear_facts` cleared all facts or just gathered facts.  The plays in question are not using gather_facts at all.  Thanks!

**Comment by ryanmerolle** [comment]
`clear_facts`seems to only clear gathered facts and not all facts. If it cleared all facts, introduced an option under `clear_facts` to set what type of facts to clear, or introduced a whole new `meta` option to clear all facts, this could give users a new set of tools to keep resource usage and play runtime down.

**Comment by bcoca** [comment]
`clear_facts` clears 'facts' .. `set_facts` does NOT create 'facts' by default (only when using `cacheable=true`), even when it does it actually creates 2 different variables scopes with the same value, a 'fact' and a hostvar (what set_fact creates by default). So you are safe using it this way.

 If you were using `vars` it would not be safe since 'lazy evaluation' would require the original variables to be available to template on each use.

**Comment by ryanmerolle** [comment]
Good insight @bcoca .

Thanks

**Comment by vadirajks** [comment]
ansible is taking too much time than normal execution. please help me if anything i need to change to improve this time (0m5.368s)
$ cat hosts 
[testhost]
10.100.64.12

$ time ssh 10.100.64.12 "date"
Sun Apr 17 11:06:10 UTC 2022
real	0m2.639s
user	0m0.015s
sys	0m0.005s

$ time ansible -i hosts testhost -m shell -a "date"
10.100.64.12 | CHANGED | rc=0 >>
Sun Apr 17 11:06:36 UTC 2022
real	0m5.368s
user	0m1.107s
sys	0m0.254s

$ ansible  --version
ansible [core 2.12.4]
  config file = /home/test/ansible_deployment/ansible.cfg
  configured module search path = ['/home/test/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /home/test/.conda/envs/ansible-env-p38/lib/python3.8/site-packages/ansible
  ansible collection location = /home/test/.ansible/collections:/usr/share/ansible/collections
  executable location = /home/test/.conda/envs/ansible-env-p38/bin/ansible
  python version = 3.8.0 (default, Nov  6 2019, 21:49:08) [GCC 7.3.0]
  jinja version = 2.11.2
  libyaml = True

**Comment by Vladimir-csp** [comment]
> After completing `role_1`, ansible seemingly hangs. Debug reveals that it starts very slowly to run iterations of get_vars.
> At the end of role_1 messages were printed to the log for each host `META: role_complete for hostname`, after that a large amount of empty `META:` messages started pouring.

I found what that silent var crunching is. Ansible silently processes vars for tasks that will not run due to not being matched by `--tags` (times undeduplicated task list).

**Comment by bcoca** [comment]
since 'tags' can be dynamic (set to a variable) ... we don't know until they are processed and templated, it could be different it tags were 'static'

**Comment by fungusakafungus** [comment]
I noticed that in my setup(`fact_caching = json`), a lot of time is spent deserializing json, so I changed it to `fact_caching = pickle` and got an almost 2x improvement.

**Comment by Vladimir-csp** [comment]
> since 'tags' can be dynamic (set to a variable) ... we don't know until they are processed and templated, it could be different it tags were 'static'

Feature request: a config parameter to turn this feature off.

**Comment by bcoca** [comment]
@Vladimir-csp you are basically asking for us to code 2 diff parsing engines and use them on a toggle

**Comment by bcoca** [comment]
waiting on data tagging

**Comment by JakeCooper** [comment]
We run hundreds of hosts and the timing blow up is incredibly painful

Restarting a systemd process accounts for ~3 minutes of time

![CleanShot 2023-12-04 at 13 55 20@2x](https://github.com/ansible/ansible/assets/5499880/a6caf260-3f79-4ad6-826a-e65f54923aaf)
